client/trino-client/src/test/java/io/trino/client/auth/external/TestExternalAuthenticator.java

```diff
@@ -134,6 +149,153 @@ public void testReAuthenticationAfterRejectingToken()
+
+        List<Future<Request>> requests = times(4,
+                () -> authenticator.authenticate(null, getUnauthorizedResponse("Bearer x_token_server=\"http://token.uri\", x_redirect_server=\"http://redirect.uri\"", firstRequest)))
+                .map(executor::submit)
+                .collect(toImmutableList());
+
+        executor.shutdown();
+        executor.awaitTermination(1, SECONDS);
+
+        ConcurrentRequestAssertion assertion = new ConcurrentRequestAssertion(requests);
+        assertion.requests().containsExactly(null, null, null);
+        assertion.firstException().hasMessage("external authentication error")
+                .isInstanceOf(ClientException.class);
+
+        assertThat(redirectHandler.getRedirectionCount()).isEqualTo(2);
+    }
+
+    @Test(timeOut = 2000)
+    public void testAuthenticationFromMultipleThreadsWithCachedTokenAfterAuthenticateTimesOut()
+            throws Exception
+    {
+        ExecutorService executor = newFixedThreadPool(5);
+        MockRedirectHandler redirectHandler = new MockRedirectHandler()
+                .sleepOnRedirect(Duration.ofMillis(5));
+
+        ExternalAuthenticator authenticator = new ExternalAuthenticator(redirectHandler, (uri, duration) -> TokenPollResult.pending(uri), KnownToken.memoryCached(), Duration.ofMillis(1));
+        List<Future<Request>> requests = times(4,
+                () -> authenticator.authenticate(null, getUnauthorizedResponse("Bearer x_token_server=\"http://token.uri\", x_redirect_server=\"http://redirect.uri\"")))
+                .map(executor::submit)
+                .collect(toImmutableList());
+
+        executor.shutdown();
+        executor.awaitTermination(1, SECONDS);
+
+        new ConcurrentRequestAssertion(requests)
+                .requests().containsExactly(null, null, null, null);
+        assertThat(redirectHandler.getRedirectionCount()).isEqualTo(1);
+    }
+
+    @Test(timeOut = 2000)
+    public void testAuthenticationFromMultipleThreadsWithCachedTokenAfterAuthenticateIsInterrupted()
+            throws Exception
+    {
+        ExecutorService executor = newFixedThreadPool(5);
+        ExecutorService interruptableThreadPool = newFixedThreadPool(1);
+        MockRedirectHandler redirectHandler = new MockRedirectHandler()
+                .sleepOnRedirect(Duration.ofMinutes(1));
+
+        ExternalAuthenticator authenticator = new ExternalAuthenticator(redirectHandler, (uri, duration) -> TokenPollResult.pending(uri), KnownToken.memoryCached(), Duration.ofMillis(1));
+        Future<Request> interruptedAuthentication = interruptableThreadPool.submit(
+                () -> authenticator.authenticate(null, getUnauthorizedResponse("Bearer x_token_server=\"http://token.uri\", x_redirect_server=\"http://redirect.uri\"")));
+        List<Future<Request>> requests = times(2,
```

Comment:
I think you need to sleep before you schedule anything here, to make sure `interruptedAuthentication` went sleep